torii/src/utils.rs

```diff
@@ -37,48 +24,73 @@ impl<T: Encode + Send> Reply for Scale<T> {
-    }
-}
-
 pub mod body {
-    use iroha_version::error::Error as VersionError;
+    use axum::{
+        async_trait,
+        body::Bytes,
+        extract::{FromRequest, FromRequestParts, Query, Request},
+    };
+    use iroha_data_model::query::cursor::ForwardCursor;
 
     use super::*;
 
-    /// Decode body as versioned scale codec
-    pub fn versioned<T: DecodeVersioned>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy
+    /// Extractor of scale encoded versioned data from body
+    #[derive(Clone, Copy, Debug)]
+    pub struct ScaleVersioned<T>(pub T);
+
+    #[async_trait]
+    impl<S, T> FromRequest<S> for ScaleVersioned<T>
+    where
+        Bytes: FromRequest<S>,
+        S: Send + Sync,
+        T: DecodeVersioned,
     {
-        warp::body::bytes().and_then(|body: Bytes| async move {
-            T::decode_all_versioned(body.as_ref()).map_err(warp::reject::custom)
-        })
-    }
+        type Rejection = Response;
 
-    /// Recover from failure in `versioned`
-    pub fn recover_versioned(rejection: Rejection) -> Result<impl Reply, Rejection> {
-        rejection
-            .find::<VersionError>()
-            .map(warp::Reply::into_response)
-            .ok_or(rejection)
+        async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection> {
+            let body = Bytes::from_request(req, state)
+                .await
+                .map_err(IntoResponse::into_response)?;
+
+            T::decode_all_versioned(&body)
+                .map(ScaleVersioned)
+                .map_err(|err| {
+                    (
+                        axum::http::StatusCode::BAD_REQUEST,
+                        format!("Transaction Rejected (Malformed), Reason : '{err}'"),
+                    )
+                        .into_response()
+                })
+        }
     }
-}
 
-/// Warp result response type
-pub struct WarpResult<O, E>(pub Result<O, E>);
+    /// Extractor for [`ClientQueryRequest`]
+    ///
+    /// First try to deserialize body as [`SignedQuery`] if fail try to parse query parameters for [`ForwardCursor`] values
+    #[derive(Clone, Debug)]
+    pub struct ClientQueryRequestExtractor(pub ClientQueryRequest);
+
+    #[async_trait]
+    impl<S> FromRequest<S> for ClientQueryRequestExtractor
+    where
+        Bytes: FromRequest<S>,
+        S: Send + Sync,
+    {
+        type Rejection = Response;
 
-impl<O: Reply, E: Reply> Reply for WarpResult<O, E> {
-    fn into_response(self) -> warp::reply::Response {
-        match self {
-            Self(Ok(ok)) => ok.into_response(),
-            Self(Err(err)) => err.into_response(),
+        async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection> {
+            let (mut parts, body) = req.into_parts();
+            let cursor = Query::<ForwardCursor>::from_request_parts(&mut parts, &state)
+                .await
+                .map(|Query(cursor)| ClientQueryRequest::cursor(cursor));
```

Comment:
I don't think it's a right thing to extract cursor from query params, since anyone can fetch cursor.
So it should be signed and passed as body.

Reply:
Created https://github.com/hyperledger/iroha/issues/4720